pennylane/estimator/compact_hamiltonian.py

```diff
@@ -178,3 +180,306 @@ def __post_init__(self):
+        if self._commuting_groups is not None:
+            hashable_commuting_groups = tuple(
+                _sort_and_freeze(group) for group in self._commuting_groups
+            )
+        elif self._pauli_dist is not None:
+            hashable_commuting_groups = _sort_and_freeze(self._pauli_dist)
+        else:
+            hashable_commuting_groups = None
+
+        hashable_params = (
+            self._num_qubits,
+            self._num_pauli_words,
+            self._max_weight,
+            hashable_commuting_groups,
+            self._one_norm,
+        )
+        return hash(hashable_params)
+
+    @property
+    def num_qubits(self):
+        """The number of qubits the Hamiltonian acts on"""
+        return self._num_qubits
+
+    @property
+    def num_pauli_words(self):
+        """The number of Pauli words (or terms) in the sum."""
+        return self._num_pauli_words
+
+    @property
+    def max_weight(self):
+        """The maximum number of Pauli operators (tensored) in any given term in the sum."""
+        return self._max_weight
+
+    @property
+    def one_norm(self):
+        """The one-norm of the Hamiltonian."""
+        return self._one_norm
+
+    @property
+    def pauli_dist(self):
+        """A dictionary representing the distribution of Pauli words in the Hamiltonian"""
+        return copy.deepcopy(self._pauli_dist)
+
+    @property
+    def commuting_groups(self):
+        """A list of groups where each group is a distribution of pauli words such that each
+        term in the group commutes with every other term in the group."""
+        return copy.deepcopy(self._commuting_groups)
+
+
+def _sort_and_freeze(pauli_dist: dict) -> tuple[tuple]:
+    """Map a dictionary into a sorted and hashable tuple"""
+    return tuple((k, pauli_dist[k]) for k in sorted(pauli_dist))
+
+
+def _validate_pauli_dist(pauli_dist: dict) -> bool:
+    """Validate that the pauli_dist is formatted as expected"""
+    for pauli_word, freq in pauli_dist.items():
+        if (not isinstance(pauli_word, str)) or (
+            not all(char in {"X", "Y", "Z"} for char in pauli_word)
+        ):
+            raise ValueError(
+                f"The keys represent Pauli words and should be strings containing either 'X','Y' or 'Z' characters only. Got {pauli_word} : {freq}"
+            )
+
+        if not (isinstance(freq, int) and (not isinstance(freq, bool)) and (freq >= 0)):
+            raise ValueError(
+                f"The values represent frequencies and should be positive integers, got {pauli_word} : {freq}"
+            )
+
+
+def _pauli_dist_from_commuting_groups(commuting_groups: tuple[dict]):
+    """Construct the total Pauli word distribution from the commuting groups."""
+    total_pauli_dist = defaultdict(int)
+
+    for group in commuting_groups:
+        for pauli_word, frequency in group.items():
+            total_pauli_dist[pauli_word] += frequency
+
+    return total_pauli_dist
+
+
+def _validate_inputs(
```

Comment:
NIT: `_validate_inputs` implies a return type of `None` i.e. a void function rather than a function that returns a `tuple`. I would prefer a different name.

Reply:
Renamed to `_preprocess_inputs`